packages/augur-core/src/contracts/para/AMMExchange.sol

```diff
@@ -1,243 +1,325 @@
-			_noFromUser = _userNo;
-			_yesFromUser = _yesToPool + _setsToSell;
+		if (_invalidShares > _setsToSell) {
+			// We have excess Invalid shares that the user will just keep.
+			_invalidFromUser = _setsToSell;
 		} else {
-			uint256 _yesToUser = _setsToSell.sub(_userYes);
-			uint256 _noToPool = _poolConstant.div(_poolYes.sub(_yesToUser)).sub(_poolNo);
-			require(_noToPool <= _userNo.sub(_setsToSell), "AugurCP: You don't have enough NO tokens to close out for this amount.");
-			_yesFromUser = _userYes;
-			_noFromUser = _noToPool + _setsToSell;
+			// We don't have enough Invalid to actually close out the Yes/No shares. They will be kept by the user.
+			// Need to actually receive yes or no shares here since we are swapping to get partial complete sets but dont have enough yes/no to make full complete sets
+			if (_yesShares > _noShares) {
+				uint256 _noSharesToBuy = _setsToSell.sub(_noShares);
+				shareToken.unsafeTransferFrom(address(this), msg.sender, NO, _noSharesToBuy);
+				_noFromUser = _invalidFromUser;
+				_yesFromUser = _yesShares.sub(_noShares.add(_noSharesToBuy).sub(_noFromUser));
+			} else {
+				uint256 _yesSharesToBuy = _setsToSell.sub(_yesShares);
+				shareToken.unsafeTransferFrom(address(this), msg.sender, YES, _yesSharesToBuy);
+				_yesFromUser = _invalidFromUser;
+				_noFromUser = _noShares.sub(_yesShares.add(_yesSharesToBuy).sub(_yesFromUser));
+			}
+			_setsToSell = _invalidFromUser;
 		}
 
-		// materialize the complete set sale for cash
+		uint256 _cashPayout = _setsToSell.mul(numTicks);
+		require(_cashPayout >= _minCashPayout, "Proceeds were less than the required payout");
+
 		shareTransfer(msg.sender, address(this), _invalidFromUser, _noFromUser, _yesFromUser);
-		cash.transfer(msg.sender, _cashToBuy);
+        cash.transfer(msg.sender, _cashPayout);
 	}
 
-	function swap(uint256 _inputShares, bool _inputYes) external returns (uint256) {
-		(uint256 _poolNo, uint256 _poolYes) = yesNoShareBalances(address(this));
-		uint256 _poolConstant = _poolYes.mul(_poolNo);
-		if (_inputYes) {
-			uint256 _yesFromUser = _inputShares;
-			uint256 _noToUser = _poolNo.sub(_poolConstant.div(_poolYes.add(_yesFromUser)));
-			shareToken.unsafeTransferFrom(msg.sender, address(this), YES, _yesFromUser);
-			shareToken.unsafeTransferFrom(address(this), msg.sender, NO, _noToUser);
-			return _noToUser;
-		} else {
-			uint256 _noFromUser = _inputShares;
-			uint256 _yesToUser = _poolYes.sub(_poolConstant.div(_poolNo.add(_noFromUser)));
-			shareToken.unsafeTransferFrom(msg.sender, address(this), NO, _noFromUser);
-			shareToken.unsafeTransferFrom(address(this), msg.sender, YES, _yesToUser);
-			return _yesToUser;
-		}
+	function quadratic(int256 _a, int256 _b, int256 _c, uint256 _maximum) public returns (uint256) {
+		int256 _piece = SafeMathInt256.sqrt(_b*_b - (_a.mul(_b).mul(4)));
+		int256 _resultPlus = (-_b + _piece) / (2 * _a);
+		int256 _resultMinus = (-_b - _piece) / (2 * _a);
```

Comment:
```suggestion
		int256 _piece = SafeMathInt256.sqrt(_b*_b + (_a.mul(_b).mul(4)));
		int256 _resultPlus = (_b + _piece) / (2 * _a);
		int256 _resultMinus = (_b - _piece) / (2 * _a);
```
Per suggestion above, I think by passing in positive values as `b` instead of negative values, we can not only make this function `uint256` instead of `int256`, but we get rid of a bunch of awkward negatives as well.